packages/abi-utils/lib/arbitrary.ts

```diff
@@ -0,0 +1,315 @@
+import * as fc from "fast-check";
+import faker from "faker";
+import { camelCase, pascalCase } from "change-case";
+
+import * as Types from "./types";
+
+export const Parameter = () =>
+  fc
+    .tuple(
+      fc.record({
+        name: ParameterName()
+      }),
+      TypeRecord()
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const EventParameter = () =>
+  fc
+    .tuple(
+      fc.record({
+        name: ParameterName(),
+        indexed: fc.boolean()
+      }),
+      TypeRecord()
+    )
+    .map(records => records.reduce((a, b) => ({ ...a, ...b }), {}));
+
+export const EventEntry = () =>
+  fc.record({
+    type: fc.constant("event"),
+    name: EventName(),
+    inputs: fc.array(EventParameter(), { maxLength: 10 }).filter(inputs => {
+      if (inputs.filter(({ indexed }) => indexed).length > 3) {
+        // only up to 3 params can be indexed
+        return false;
+      }
+
+      // names that are not blank should be unique
+      const names = inputs.map(({ name }) => name).filter(name => name !== "");
+      return names.length === new Set(names).size;
+    }),
+    anonymous: fc.boolean()
+  });
+
+export const FunctionEntry = () =>
+  fc
+    .tuple(
+      fc.record(
+        {
+          type: fc.constant("function")
+        },
+        { withDeletedKeys: true }
+      ),
+      fc.record({
+        name: FunctionName(),
+        inputs: fc.array(Parameter(), { maxLength: 10 }).filter(inputs => {
+          // names that are not blank should be unique
+          const names = inputs
+            .map(({ name }) => name)
+            .filter(name => name !== "");
+          return names.length === new Set(names).size;
+        })
+      }),
+      fc.record(
+        {
+          outputs: fc.array(Parameter(), { maxLength: 10 }).filter(outputs => {
+            // names that are not blank should be unique
+            const names = outputs
+              .map(({ name }) => name)
+              .filter(name => name !== "");
+            return names.length === new Set(names).size;
+          })
+        },
+        { withDeletedKeys: true }
```

Comment:
I'm pretty sure `outputs` is not optional, and you don't have it as optional in your TS type; why is this here?

Reply:
Our ABI schema says it's optional... I'm not sure, maybe it used to be?

Reply:
Also note: the TS types are normalized, so they wouldn't have optionality. Maybe the thing to do here is make sure that `normalize` includes `outputs`

Reply:
Hm.  Should we change the ABI schema to have it mandatory?

Reply:
Ah, apparently not: Solidity v0.5.0 [specifies](https://solidity.readthedocs.io/en/v0.5.0/abi-spec.html#json) that `outputs` "can be omitted if function doesn’t return anything"

So seems like this should get added to `normalize`

Reply:
(which I've now done)

Reply:
Oh wow!  I didn't realize that.  Thanks for correcting that!